src/rpc/AppRpcProvider.ts

```diff
@@ -0,0 +1,156 @@
+import { Network } from '@ethersproject/networks'
+import { JsonRpcProvider, Provider } from '@ethersproject/providers'
+
+function now() {
+  return new Date().getTime()
+}
+
+function checkNetworks(networks: Array<Network>): Network | null {
+  let result: Network | null = null
+
+  for (let i = 0; i < networks.length; i++) {
+    const network = networks[i]
+
+    // Null! We do not know our network; bail.
+    if (network == null) {
+      return null
+    }
+
+    if (result) {
+      // Make sure the network matches the previous networks
+      if (
+        !(
+          result.name === network.name &&
+          result.chainId === network.chainId &&
+          (result.ensAddress === network.ensAddress || (result.ensAddress == null && network.ensAddress == null))
+        )
+      ) {
+        return null
+      }
+    } else {
+      result = network
+    }
+  }
+
+  return result
+}
+
+interface ProviderPerformance {
+  latency: number
+  failureRate: number
+  lastEvaluated: number
+}
+
+interface FallbackProviderEvaluation {
+  provider: JsonRpcProvider
+  performance: ProviderPerformance
+}
+
+/**
+ * AppRpcProvider is an extension of the JsonRpcProvider class from the ethers.js library.
+ * This provider balances requests among multiple JSON-RPC endpoints.
+ */
+export default class AppRpcProvider extends JsonRpcProvider {
+  readonly providerEvaluations: ReadonlyArray<FallbackProviderEvaluation>
+  readonly evaluationInterval: number
+
+  constructor(providers: JsonRpcProvider[], evaluationInterval = 30000) {
+    if (providers.length === 0) throw new Error('providers array empty')
+    providers.forEach((provider, i) => {
+      if (!Provider.isProvider(provider)) throw new Error(`invalid provider ${i}`)
```

Comment:
do we want to throw here without catching it ?  or do we expect to catch the error at the callsite?
we could also filter the list to just the valid providers.

Reply:
https://github.com/ethers-io/ethers.js/blob/master/packages/providers/src.ts/fallback-provider.ts

For our use case here this is probably not necessary, but I approached the task with somewhat of a mind towards publishing this later.